crates/wasmtime/src/runtime/vm/instance/allocator/pooling/metrics.rs

```diff
@@ -0,0 +1,114 @@
+use core::sync::atomic::Ordering;
+
+use crate::{Engine, vm::PoolingInstanceAllocator};
+
+/// `PoolingAllocatorMetrics` provides access to runtime metrics of a pooling
+/// allocator configured with [`crate::InstanceAllocationStrategy::Pooling`].
+///
+/// This is a cheap cloneable handle which can be obtained with
+/// [`Engine::pooling_allocator_metrics`].
+#[derive(Clone)]
+pub struct PoolingAllocatorMetrics {
+    engine: Engine,
+}
+
+impl PoolingAllocatorMetrics {
+    pub(crate) fn new(engine: &Engine) -> Option<Self> {
+        engine.allocator().as_pooling().map(|_| Self {
+            engine: engine.clone(),
+        })
+    }
+
+    /// Returns the number of core (module) instances currently allocated.
+    pub fn core_instances(&self) -> u64 {
+        self.allocator().live_core_instances.load(Ordering::Acquire)
```

Comment:
Similar to above, I'd say that using `Relaxed` here is fine since these aren't actually synchronizing with anything other than themselves.